bigquery/storage/managedwriter/client.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package managedwriter
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+	"strings"
+
+	"cloud.google.com/go/bigquery"
+	storage "cloud.google.com/go/bigquery/storage/apiv1beta2"
+	"google.golang.org/api/option"
+	storagepb "google.golang.org/genproto/googleapis/cloud/bigquery/storage/v1beta2"
+)
+
+// Client is a managed BigQuery Storage write client scoped to a single project.
+type Client struct {
+	rawClient *storage.BigQueryWriteClient
+	projectID string
+}
+
+// NewClient instantiates a new client.
+func NewClient(ctx context.Context, projectID string, opts ...option.ClientOption) (c *Client, err error) {
+	numConns := runtime.GOMAXPROCS(0)
+	if numConns > 4 {
+		numConns = 4
+	}
+	o := []option.ClientOption{
+		option.WithGRPCConnectionPool(numConns),
+	}
+	o = append(o, opts...)
+
+	rawClient, err := storage.NewBigQueryWriteClient(ctx, o...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		rawClient: rawClient,
+		projectID: projectID,
+	}, nil
+}
+
+// NewManagedStream establishes a new stream for writing.
+func (c *Client) NewManagedStream(ctx context.Context, table *bigquery.Table, opts ...WriterOption) (*ManagedStream, error) {
```

Comment:
I think table could just be datasetID and tableID. I worry about this client reaching for code in both directions up and down the file structure. This means the core bigquery package could never rely on this package due to cycles.

Reply:
Fine switching it to string, now I'm wondering if it should be a variadic argument.  We don't do anything with the reference other than for the construct a stream case, but if we allow users to pass in an optional stream ID perhaps we let them treat table as optional as well?  Will switch it and augment the validation (your other comment) and see how it plays.